osquery/tables/system/windows/registry.cpp

```diff
@@ -266,33 +269,145 @@ void queryKey(const std::string& keyPath, QueryData& results) {
   }
   delete[](bpDataBuff);
   RegCloseKey(hRegistryHandle);
-};
+}
 
-QueryData genRegistry(QueryContext& context) {
-  QueryData results;
-  std::set<std::string> rKeys;
-  auto shouldWarnLocalUsers = false;
-  /// By default, we display all HIVEs
-  if ((context.constraints["key"].exists(EQUALS) &&
-       context.constraints["key"].getAll(EQUALS).size() > 0)) {
-    rKeys = context.constraints["key"].getAll(EQUALS);
-    shouldWarnLocalUsers = true;
-  } else {
-    for (auto& h : kRegistryHives) {
-      rKeys.insert(h.first);
+void populateDefaultKeys(std::set<std::string>& rKeys) {
+  for (const auto& hive : kRegistryHives) {
+    rKeys.insert(hive.first);
+  }
+}
+
+void replaceKeysWithSubkeys(std::set<std::string>& rKeys) {
+  std::set<std::string> newKeys{};
+  for (const auto& key : rKeys) {
+    QueryData regResults;
+    queryKey(key, regResults);
+    for (const auto& r : regResults) {
+      if (r.at("type") == "subkey") {
+        newKeys.insert(r.at("path"));
+      }
+    }
+  }
+  rKeys = newKeys;
+}
+
+void appendSubkeyToKeys(const std::string& subkey,
+                        std::set<std::string>& rKeys) {
+  std::set<std::string> newKeys{};
+  for (auto& key : rKeys) {
+    newKeys.insert(key + kRegSep + subkey);
+  }
+  rKeys = newKeys;
+}
+
+Status populateAllKeysRecursive(std::set<std::string>& rKeys,
+                                int currDepth,
+                                int maxDepth) {
+  std::set<std::string> subkeys{};
+
+  if (currDepth > maxDepth) {
+    return Status(
+        1, "Max recursive depth (" + std::to_string(maxDepth) + ") reached");
+  }
+
+  for (const auto& key : rKeys) {
+    QueryData regResults;
+    queryKey(key, regResults);
+    for (const auto& r : regResults) {
+      if (r.at("type") == "subkey") {
+        subkeys.insert(r.at("path"));
+      }
+    }
+  }
+
+  if (!(subkeys.size() == 0)) {
+    auto status = populateAllKeysRecursive(subkeys, ++currDepth);
+    if (!status.ok()) {
+      return status;
+    }
+    rKeys.insert(subkeys.begin(), subkeys.end());
+  }
+
+  return Status(0, "OK");
+}
+
+Status resolveRegistryGlobs(const std::string& pattern,
+                            std::set<std::string>& results) {
+  auto pathElems = osquery::split(pattern, kRegSep);
```

Comment:
I had assumed it would always return at least '', but that is a pretty bad assumption.